### ISSUE TITLE
Feature Request: Access to channel/session objects from within a request handler

### ISSUE THREAD

**Issue by colstrom** [hirura/hrr_rb_ssh#5]
Narrowly, I would like to be able to close a channel immediately from within a request handler in certain situations.

More generally, it would be nice to have access to the session/channel objects.

As a workaround, I'm doing this at the moment...

```ruby
::ObjectSpace.each_object.select do |object|
  ::HrrRbSsh::Connection::Channel === object
end.find do |channel|
  ::Thread.current === ::Kernel
                         .instance_method(:instance_variable_get)
                         .bind(channel)
                         .call("@channel_loop_thread")
end.close
```

... but that's a very large hammer.

Would you be open to exposing session/channel to handlers?

**Comment by hirura** [comment]
@colstrom,
I thought that finishing `context.chain_proc` is sufficient for closing session-type channel purpose.

```ruby
  HrrRbSsh::Connection::RequestHandler.new { |context|
    context.chain_proc { |chain|
      begin
        loop do
          buf = context.io[0].readpartial(10240)
          some_validation(buf) # raise if validation failure
          ret = some_method(buf)
          context.io[1].write ret
        end
        exitstatus = 0
      rescue => e
        exitstatus = 1
      end
    }
  }
```

Could you let me know in what case channel or session need to be closed so that I understand what to be improved ?

**Comment by colstrom** [comment]
In my scenario, `request_shell` is handled by advising the client that the service does not offer a shell, printing some usage help, and then explicitly closing the channel.

It's totally possible that I don't understand what guarantees `chain_proc` provides, and why/how to use it correctly. It seemed optional, from my read of the documentation, but without it, the connection does not seem to terminate. If it is intended that `chain_proc` be optional, then having a way to close the connection would be welcome.

Happy to provide any additional information that could help clarify things.

**Comment by hirura** [comment]
@colstrom,

I’ll try to explain how to handle channels with using `proc_chain`. Please note that the code blocks below are just for explanation. They would not work as they are. Please see [reference request handlers](https://github.com/hirura/hrr_rb_ssh/tree/master/lib/hrr_rb_ssh/connection/request_handler) for working examples.

When an SSH server receives a channel-request, the server must returns channel-success or channel-failure message to the client. On the other hand, input and output data must be able to be sent or received at any time. And the data stream must be able to be closed from server side. To address the behaviour, I implemented the `proc_chain`, which is running in independent thread, to separate it from channel-request and its response.

For example, an SSH client that requests interactive shell has the sequence: pry-req, then shell. When a pry-req channel-request comes and the callback implementation is the following,

```
conn_pty_req = HrrRbSsh::Connection::RequestHandler.new { |context|
  ptm, pty = get_pty()
  context.vars[:ptm], context.vars[:pty] = ptm, pty # context.vars is used to pass values to following requests
  exitstatus = 0
}
```

Then `conn_pty_req` callback implementation will return success or failure immediately (returning `0` means success) and `pty` and `ptm` are kept for the following requests.

Then the SSH client sends shell channel-request. Assume that shell implementation is the following.

```
conn_shell = HrrRbSsh::Connection::RequestHandler.new { |context|
  context.io[1].write ‘some message’
  exitstatus=0
}
```

Of course, this works for one-time messaging. The server returns success message to the shell channel-request, and then (or before the channel-request response) the server sends `’some message’` to the client.

Here, I think, the reason you wanted to use a thread is to handle data stream in background or independently from channel-requests like the following code.

```
conn_shell = HrrRbSsh::Connection::RequestHandler.new { |context|
  Thread.new {
    context.io[1].write context.io[0].read() # this is an example echo server
  }
  exitstatus=0
}
```

This would work as expected except for closing the channel, which situation you got.

Instead of writing your own thread, it is possible to use `chan_proc`. The purpose of the `chain_proc` is to handle data stream independently from channel-requests and to close channel safely.

When a channel is opened, the SSH server implementation creates a proc chain thread and the thread waits chained procs implemented as `proc_chain`.

```
conn_shell = HrrRbSsh::Connection::RequestHandler.new { |context|
  context.chain_proc { |chain|
    loop {
      context.io[1].write context.io[0].read() # expect read() to raise error to break loop when io is closed
    }
  }
  exitstatus=0
}
```

When `chain_proc` method is called in channel-request, the code block is queued in proc chain, and called in the proc chain thread. And when the proc is finished, the underlying channel is closed. Actually the server sends EOF to the client.

The chain proc , as it is named, can be written like chain. For example, in pry-req channel-request, the `pty` and `ptm` can be closed safely.

```
conn_pty_req = HrrRbSsh::Connection::RequestHandler.new { |context|
  ptm, pty = get_pty()
  context.vars[:ptm], context.vars[:pty] = ptm, pty # context.vars is used to pass values to following requests
  context.chain_proc{ |chain|
    begin
      chain.next # calls a chain_proc block in subsequent channel-request
    ensure
      ptm.close; pty.close # safely close after chain_procs are finished
    end
  }
  exitstatus = 0
}
```

In the above code, `chain.next` calls subsequent `chain_proc` code. When `chain.next` is called in a running chain proc thread, the thread waits a next `chain_proc` code is called in subsequent channel-request.

I hope this can help you.

**Comment by colstrom** [comment]
Thanks for the detailed explanation!

I had reviewed the reference implementations earlier, to get a better understanding of the library. They were quite helpful, so thanks for providing those.

In my case, I'm not creating any new threads. If I implied that I was, my apologies for that. Here's two examples that may better illustrate the situation:

```ruby
class HangingRequestHandler
  def self.run(context)
    context.io[1].puts "message"
  end
end
```

```ruby
class ReasonableRequestHandler
  def self.run(context)
    context.chain_proc do
      context.io[1].puts "message"
    end
  end
end
```

With a server like so:

```ruby
class Server
  OPTIONS = {
    "connection_channel_request_shell" => HangingRequestHandler
  }

  def initialize(address: "127.0.0.1", port: 50022)
    @server = TCPServer.new(address, port)
  end

  def listen
    loop do
      HrrRbSsh::Server.new(@server.accept, OPTIONS).start
    rescue HrrRbSsh::Error::ClosedTransport
    end
  end
end

Server.listen
```

When a client requests a shell, it will write `message`, and the connection will remain open. Neither `^C` nor `^D` from the client will end the session (which makes sense because the request handler above doesn't do anything with them).

Running that same server with

```ruby
OPTIONS["connection_channel_request_shell"] = ReasonableRequestHandler
```

results in connections closing after writing `message`. This is the expected behaviour.


Given that the only difference between `HangingRequestHandler` and `ReasonableRequestHandler` is that `ReasonableRequestHandler` uses `chain_proc`... it seems like `chain_proc` is *required*?

My impression was that it was optional, in which case it would make sense to be able to explicitly close a channel or session.


Unrelated to this (I can open a separate issue if you'd prefer), the example of passing `pts` and `ptm` via `context.variables` is interesting. Is there a way to pass things between `authenticator` contexts (and `request_handler` contexts) in a similar manner? I'm thinking of two scenarios there...

First, passing between `authenticator` contexts would allow for things like requiring *both* public key and password authentication, or other multi-step auth flows.

Secondly, passing between `authenticator` and `request_handler` would allow for things like creating an encryption context from the user's password, and using that for encrypting/decrypting things like stored secrets.

**Comment by hirura** [comment]
@colstrom,

I'm now clear regard to things what you mention. The answer is YES. In the current library implementation, `chain_proc` is required to close the underlying session channel on the server side.
Let me consider to revise instructions in README or to provide some new method like `context.close_channel`.

And regard to the context passing between authentications and between authentication and channel-request, it's very interesting. Could you please open another issue so that we can discuss about that ?

**Comment by hirura** [comment]
I have implemented `Context#close_session`. Please refer to #6 and try.

The new method can be used as follows.

```ruby
conn_shell = HrrRbSsh::Connection::RequestHandler.new { |context|
  context.io[1].write 'message'
  context.close_session
}
```

**Comment by colstrom** [comment]
Separate issue opened, as requested.

For testing #6, I'll build from source and give that a spin.

Related, have you considered publishing [prerelease gems](https://guides.rubygems.org/patterns/#prerelease-gems) to facilitate testing of things like this?

Thanks for the quick turnaround.

**Comment by colstrom** [comment]
Testing with 6fdb54b, `context.close_session` works great in a request handler without `chain_proc`.

This addresses my narrow case, thank you.

As far as the general case goes (having access to the session/channel from a request handler):
 - I don't see much utility in direct access to the Session. Closing it was the main thing, and this change allows that without exposing the Session directly.
 - Access to the Channel may still be useful, specifically for the `send_channel_*` methods, when implementing something like the [Public Key Subsystem](https://tools.ietf.org/html/rfc4819).

**Comment by hirura** [comment]
I have released [v0.3.0.pre1](https://rubygems.org/gems/hrr_rb_ssh/versions/0.3.0.pre1).

As for the access to the Channel, let me read the RFC document to study what to be done potentially.

**Comment by hirura** [comment]
I have read the Public Key Subsystem RFC document and determined NOT to expose the Channel in request handlers. This is because some functions implemented as request handler instance are running on an established SSH channel, and only the SSH channel itself uses the Channel's `send_channel_*` methods.

The Public Key Subsystem, which is mentioned, actually does not need to use SSH Channel's `send_channel_*` methods, in other words the subsystem doesn't extend or replace existing SSH messages because the data stream of the subsystem runs on SSH channel's SSH_MSG_CHANNEL_DATA message.

An implementation of such subsystem would be as follows.

```ruby
subsys = HrrRbSsh::Connection::RequestHandler.new { |context|
  context.chain_proc { |chain|
    case context.subsystem_name
    when 'SUBSYSTEM NAME'
      SUBSYS_NAME_CLASS.start_some_service(context.io)
      exitstatus = 0
    else
      raise "Unsupported subsystem: #{context.subsystem_name}"
    end
  }
}
```

On the other hand, when extending or replacing existing SSH behavior is really required, `Channel` class should be opened and monkey-patched.

**Comment by colstrom** [comment]
That sounds reasonable. I'll defer to your greater understanding of the protocol here. If it doesn't seem necessary to expose the Channel to the request handler to implement something like that, then I don't see any immediate need to do so.

If at some point I encounter a situation that warrants it, I'll open a new request at that time, with a concrete example to support it. Right now it just "seems like it could maybe be useful", which isn't really a great reason to expose it. ;p

Also, I just tried the prerelease, and it works as expected.

Thanks for taking the time to dig into the RFC, for making the prerelease available, and for implementing the practical parts of this feature request. I think everything has been addressed now, so feel free to close this issue at your convenience (I'm not sure if you prefer to close it now, or once the features make it into the main release).
